File: Data Structures & Algorithms/combinations-of-a-phone-number/submission-0.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        hashmap = {'2': ['a', 'b', 'c'], '3': ['d', 'e', 'f'], '4': ['g', 'h', 'i'], '5': ['j', 'k', 'l'], '6': ['m', 'n', 'o'], '7': ['p', 'q', 'r', 's'], '8': ['t', 'u', 'v'], '9': ['w', 'x', 'y', 'z']}
        res = []

        def backtrack(curr, i):
            if i == len(digits):
                res.append("".join(curr))
                return

            for letter in hashmap[digits[i]]:
                curr.append(letter)
                backtrack(curr, i + 1)
                curr.pop()
        
        # Step 3: handle edge case for empty input
        if not digits:
            return []

        curr = []
        backtrack(curr, 0)

        return res
            

File: Data Structures & Algorithms/longest-common-prefix/submission-0.cpp
class Solution {
public:
    string longestCommonPrefix(vector<string>& strs) {
        string res = strs[0];
        for (int i = 1; i < strs.size(); i++) {
            string curr = "";
            int len = max(res.length(), strs[i].length());
            for (int j = 0; j < len; j++) {
                if (res[j] != strs[i][j]) {
                    break;
                }

                curr += res[j];
            }

            res = curr;
        }

        return res;
    }
};